File: current_session/python/63.py
"""
63
"""
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        H, W = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[0 for i in range(W)] for j in range(H)]
        i, j = H-1, W-1
        while i >= 0 and obstacleGrid[i][-1] != 1:
            dp[i][-1] = 1
            i -= 1
        while j >= 0 and obstacleGrid[-1][j] != 1:
            dp[-1][j] = 1
            j -= 1
        
        for i in range(H-2, -1, -1):
            for j in range(W-2, -1, -1):
                if obstacleGrid[i][j] == 1:
                    dp[i][j] = 0
                else:
                    dp[i][j] = dp[i+1][j] + dp[i][j+1]
        
        return dp[0][0]

print(Solution().uniquePathsWithObstacles([[0,0,0],[0,1,0],[0,0,0]]))
print(Solution().uniquePathsWithObstacles([[0,1],[0,0]]))

File: current_session/c++/63.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
    int H = obstacleGrid.size(), W = obstacleGrid[0].size();
    if (obstacleGrid[0][0] == 1 or obstacleGrid[H-1][W-1] == 1) return 0;

    // use an 2D array => DP
    vector<vector<int>> dp(H, vector<int>(W, 0));
    dp[0][0] = 1;

    for (int i=0; i < H; ++i) {
      for (int j=0; j < W; ++j) {
        if (obstacleGrid[i][j] == 0) {
          int add = 0;
          if (i-1 >= 0 && j >= 0) add += dp[i-1][j];
          if (i >= 0 && j-1 >= 0) add += dp[i][j-1];
          dp[i][j] += add;
        }
      }
    }

    return dp[H-1][W-1];
  }
};

int main() {
  cout << Solution().uniquePathsWithObstacles() << endl;
}